### DISCUSSION TITLE
How to create and use another connection provider instead of the WebSocket one?

### DISCUSSION THREAD

**Discussion by Tnze** [TypeFox/monaco-languageclient#583] [Q&A]
Hi!
I'm trying to add LSP supports for a IDE project I'm developing based on the Tauri framework.
In Tauri, there is a better way to do the IPC instead of using json-rpc based on WebSocket connection.

So does `monaco-languageclient` supports this case? How should I do?
Should I implement the `IConnectionProvider` and pass it to constructor of `MonacoLanguageClient`, or directly use `BaseLanguageClient`? I'm confusing and unable to find out.

And what is the different between `monaco-languageclient` and `vscode-languageclient`?
Should I install `vscode-languageclient` as a dependency of my project? Is that normal a language client for monaco-editor is depending on vscode things?

Thanks for your help.

**Comment by msujew** [comment]
@Tnze `vscode-languageclient` is the package used for implementing **vscode** extensions that add support for a language to **vscode** via a language server. It acts as a language client for vscode (as the name suggests).

This project (i.e. the `monaco-languageclient` library) does the same but just for the monaco editor as a standalone component.

Installing `vscode-languageclient` is therefore not necessary.

> Should I implement the `IConnectionProvider` and pass it to constructor of `MonacoLanguageClient`

Ideally yes. The `MessageTransports` object returned by the `IConnectionProvider#get` function determines how the monaco language client communicates to the language server. If you can implement that using IPC via Tauri, then go for it.

cc @kaisalmen Can you convert this to a discussion? I don't seem to have maintainer rights on this repo.

**Comment by Tnze** [comment]
@msujew Thank you! So you means:
1. Installing `vscode-languageclient` is therefore not necessary.
2. I should implement `IConnectionProvider#get`.

But to do so, I think I should import `MessageTransports` type somewhere? As I can see it's belongs to the `vscode-languageclient`:
![image](https://github.com/TypeFox/monaco-languageclient/assets/19628575/dd01a290-415e-40cd-bfde-a3a8963fa32f)

I known its ok to avoid using `MessageTransports` type according to some duck type feature. But if it's actually what you said
> This project (i.e. the monaco-languageclient library) does the same but just for the monaco editor as a standalone component.

It's weird right?

Thank you!

**Reply by kaisalmen** [reply]
Hi @Tnze I converted this into a discussion.

This project has some history. It nowadays relies completely on `@codingame/monaco-vscode-api`, see this explanation: https://github.com/CodinGame/monaco-vscode-api#history

> Installing vscode-languageclient is therefore not necessary.

That is no correct. `monaco-languagclient` is only a thin layer above it today and therefore it is required, but is also a direct dependency that should automatically installed. `@codingame/monaco-vscode-api` provides the glue so `monaco-editor` can use an unaltered `vscode-languageclient`.

> Ideally yes. The MessageTransports object returned by the IConnectionProvider#get function determines how the monaco language client communicates to the language server. If you can implement that using IPC via Tauri, then go for it.

See how this is done in the python client example: https://github.com/TypeFox/monaco-languageclient/blob/main/packages/examples/src/python/client/main.ts#L34-L39
https://github.com/TypeFox/monaco-languageclient/blob/main/packages/examples/src/python/client/main.ts#L68-L72
Just plug in what you want to use.

**Reply by Tnze** [reply]
Thanks for answering.

I think I fully understand!

I have already successfully received jsonrpc messages at the Rust side (the backend of Tauri application).

It seems like there are only forwarding messages to the LSP server and some process management works to do, for my target. This package really helps.

**Reply by luishugoooo** [reply]
Hi, would you be okay with sharing some of your monaco/tauri/LSP experience? I am trying to create a small LaTeX editor using [this LSP](https://github.com/latex-lsp/texlab) and I feel somewhat stuck
